### PR TITLE
[SYCL] Refactor address space attributes to align with llorg.

### PR DESCRIPTION
**Pull request by bader**
This PR contains two patches:
1. [SYCL] Implement SYCL address space attributes handling - cherry-pick from llorg of https://github.com/llvm/llvm-project/commit/7818906ca134775edffb77857f436359d3a50b90.
2. Adjust DPC++ to address space handling changes from llorg - changes on top of llorg commit to keep DPC++ working.

Today use re-use OpenCL attributes in SYCL mode, so second patch updates these uses with SYCL attributes (not all of them though). 
Second patch includes content of https://reviews.llvm.org/D100396 + additional changes:
1. Enable `opencl_constant` address space in SYCL mode.
2. SYCL mode reuses a lot of OpenCL code, which rely on OpenCL language address spaces. Switching to new SYCL address spaces breaks re-use "as is". I updated uses of `opencl_local` address space, but there are a lot of uses for `opencl_global` (e.g. special types like images, pipes, etc. are annotated with OpenCL language address space). We will have to introduce SYCL address spaces in less invasive way possible.
3. Identified that https://github.com/llvm/llvm-project/commit/7818906ca134775edffb77857f436359d3a50b90 drops address space attributes for global variables. I'll upload fix for this issue to llorg shortly.
4. I left a couple of FIXME comments, which should be addressed eventually.
5. clang/lib/Sema/SPIRVBuiltins.td updated to use SYCL address spaces. @Naghasan, please, take a look. I'm not sure if SPIRVBuiltins.td is supposed to be used for SYCL language mode only.
6. Reverted a whitespace few changes and one LIT test (clang/test/SemaOpenCLCXX/address-space-lambda.clcpp) to match llorg source code.
7. Reverted 761b2ea55a1d9b9dae52031d62b26830269aa926, which was a workaround for custom SYCL mangling scheme for NVPTX target.

### CODE REVIEW

**Comment by AaronBallman** [comment]
This PR is marked as a draft -- do you want me to hold off on reviewing it until you remove that, or would you like review now?

File: clang/lib/AST/ItaniumMangle.cpp
Comment (bader) [inline clang/lib/AST/ItaniumMangle.cpp:2505]: It looks like this change remove the need for SYCL bindings in `libclc` library.
Before this change, SYCL device compiler used non-standard mangling for NVPTX target and mangled differently raw pointer parameters and parameters annotated with `opencl_private` address space attribute, although there is no need to differentiate these cases.
New code says the different mangling is used for parameters annotated with `opencl_private` address space attribute only if they are mapped to non-0 address space, so SYCL mangling for NVPTX now aligned with OpenCL mangling for NVPTX and we be remove the hack from libclc CMake files.
@Naghasan, am I right?

Comment (bader) [reply clang/lib/AST/ItaniumMangle.cpp:2505]: llvm-test-suite results seem to confirm that.
  Unsupported      : 236
  Passed           : 244
  Expectedly Failed:  25

**Comment by bader** [comment]
> This PR is marked as a draft -- do you want me to hold off on reviewing it until you remove that, or would you like review now?

I think it should be ready for review now. Hopefully CI checks won't detect new issues with this patch.

File: clang/lib/Basic/Targets/SPIR.h
Comment (premanandrao) [inline clang/lib/Basic/Targets/SPIR.h:140]:
```suggestion
    // language semantics along with the semantics of embedded C's default
```

File: clang/test/CodeGenSYCL/address-space-conversions.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/address-space-conversions.cpp:1]: Please add a comment here about what this test tests.

File: clang/test/CodeGenSYCL/address-space-deduction.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/address-space-deduction.cpp:2]: Please add a comment here about what this test tests.

File: clang/test/CodeGenSYCL/address-space-deduction.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/address-space-deduction.cpp:11]: This alloca is for the array below, right?

File: clang/test/SemaSYCL/address-space-conversions.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/address-space-conversions.cpp:1]: Please add a comment here about what the test tests.

**Review by premanandrao — COMMENTED** [review]
@bader, I have tried to best review the changes and they mostly look good, but I do not understand it fully.    I just have a few nits and some minor questions.

File: clang/include/clang/Sema/ParsedAttr.h
Comment (AaronBallman) [inline clang/include/clang/Sema/ParsedAttr.h:663]: If this is returning SYCL address spaces, should this be returning `opencl_constant`?

File: clang/lib/Basic/Targets/SPIR.h
Comment (AaronBallman) [inline clang/lib/Basic/Targets/SPIR.h:138]: Are the SYCL spec maintainers aware of this implementation issue?

File: clang/lib/CodeGen/CodeGenModule.cpp
Comment (AaronBallman) [inline clang/lib/CodeGen/CodeGenModule.cpp:271]: *thank you for getting rid of that stupid `llvm_unreachable`!*

This closes: https://github.com/intel/llvm/issues/3622

File: clang/lib/Basic/Targets/SPIR.h
Comment (Naghasan) [reply clang/lib/Basic/Targets/SPIR.h:138]: This comment is super strange, AMDGPU target is doing this exact thing (default is generic or flat address space) with a view on killing the map with private as default (see AMDGPU's `adjust` comment). NVPTX doesn't because the generic and private (local in the CUDA terminology) maps to the same address space, however the semantic used for Default matches the semantic of a flat address space.

File: clang/lib/Sema/SPIRVBuiltins.td
Comment (Naghasan) [inline clang/lib/Sema/SPIRVBuiltins.td:38]: to answer your question @bader, this was meant to be used in both OpenCL C and SYCL. However, I think your change has no effect as this line https://github.com/intel/llvm/pull/3634/files#diff-e0ad0cd86168ae6b1bdbc14d96dce7bd4de45ec1da69b9ac18dacb7adf1e453cR6520 is doing the conversion based on the language opts.

File: clang/include/clang/Sema/ParsedAttr.h
Comment (bader) [reply clang/include/clang/Sema/ParsedAttr.h:674]: Yes.

File: clang/include/clang/Sema/ParsedAttr.h
Comment (bader) [reply clang/include/clang/Sema/ParsedAttr.h:676]: > It looks odd to radically change to `Default`, this could potentially have side effects for the host code (like overload redefinition).

Is this really a **change**? I think `Default` address space is used by default, so I don't expect a "radical change".
NOTE: OpenCL is using the same mapping.

> 
> I would suggest to keep unchanged the address space if not in the handled cases.

I'm not sure what do you mean. This function maps `ParsedAttr` to `LangAs` enum. According to my understanding using `Default` address space by default achieves the described behavior.

File: clang/lib/Basic/Targets/SPIR.h
Comment (bader) [reply clang/lib/Basic/Targets/SPIR.h:138]: I would be happy to avoid these SYCL customizations, but I don't think we can do better w/o breaking SPIR ABI.
For instance, SPIR-V translator relies on mapping address space `0` to `private` address space (in OpenCL terms).
Let me know if you have other ideas.

File: clang/test/CodeGenSYCL/address-space-deduction.cpp
Comment (bader) [reply clang/test/CodeGenSYCL/address-space-deduction.cpp:11]: Right. We have to put checks here because CodeGen puts alloca+addrspaecast instructions at the beginning of the function.

File: clang/test/CodeGenSYCL/address-space-conversions.cpp
Comment (bader) [reply clang/test/CodeGenSYCL/address-space-conversions.cpp:1]: Done.

File: clang/test/CodeGenSYCL/address-space-deduction.cpp
Comment (bader) [reply clang/test/CodeGenSYCL/address-space-deduction.cpp:2]: Done.

File: clang/test/SemaSYCL/address-space-conversions.cpp
Comment (bader) [reply clang/test/SemaSYCL/address-space-conversions.cpp:1]: Done.

File: clang/lib/Basic/Targets/SPIR.h
Comment (Naghasan) [reply clang/lib/Basic/Targets/SPIR.h:138]: Not sure to whom you are answering :) but

> I would be happy to avoid these SYCL customizations, but I don't think we can do better w/o breaking SPIR ABI.

That was the underlying reason of my comment: I don't think you can avoid this but this is not unprecedented anyway. Because of that, this comment is strange.

File: clang/include/clang/AST/Type.h
Comment (Naghasan) [reply clang/include/clang/AST/Type.h:489]: :+1:

File: clang/include/clang/Sema/ParsedAttr.h
Comment (Naghasan) [reply clang/include/clang/Sema/ParsedAttr.h:678]: > I'm not sure what do you mean. This function maps ParsedAttr to LangAs enum.

Oups, I completely miss-read this. For some reason, I got stuck into "this function does LangAs to LangAs"...

So, the function LGTM :)

File: clang/lib/Basic/Targets/SPIR.h
Comment (bader) [reply clang/lib/Basic/Targets/SPIR.h:138]: This comment was suggested by Anastasia here: https://reviews.llvm.org/D89909#inline-946388, https://reviews.llvm.org/D89909#inline-948325.
I'm open to improvements. Do you suggest removing "FIXME:" word from the comment?

**Review by premanandrao — DISMISSED** [review]
Thanks for addressing my review comments.   I am okay with this change.

**Comment by bader** [comment]
Thanks!
@AaronBallman, @Naghasan, does it look good to you?

File: clang/lib/Basic/Targets/SPIR.h
Comment (AaronBallman) [reply clang/lib/Basic/Targets/SPIR.h:138]: If I'm understanding properly, it sounds like the FIXME is wrong -- the code doesn't need to be fixed, it's basically behaving as the implementation requires it, and what needs to change is the SYCL spec based on implementation feedback. So perhaps change if it from a FIXME to a NOTE and file an issue with the SYCL spec authors?

**Review by AaronBallman — DISMISSED** [review]
LGTM aside from the commenting suggestion.

File: clang/lib/Basic/Targets/SPIR.h
Comment (bader) [reply clang/lib/Basic/Targets/SPIR.h]: Changed FIXME to NOTE.

I'm not sure if there is an issue with the SYCL spec. The problem we discussed https://reviews.llvm.org/D89909 is that we now have two different mappings from AST/language address space to LLVM address space. The only difference between these is how `Default` address space is mapped. I see two different points of view on how `Default` should be treated for GPU programming languages.  Anastasia thinks `Default` should be mapped to OpenCL `private` address space, but this logic doesn't work in case of SYCL and other language mode, which do not infer address space attribute at AST level. SYCL spec authors are well aware of this issue and I think SYCL 2020 spec has clarifications for this.

@Naghasan, feel free to correct me.

**Comment by bader** [comment]
@kbobrovs, @DenisBakhvalov, I see failures in ESIMD/private_memory/pm_access_2.cpp and ESIMD/private_memory/pm_access_3.cpp after [comment update](https://github.com/intel/llvm/pull/3634/commits/f57217b002479646bef4e732d4f24cb8dd9a4e12):

> warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi1EEEE allocation size is too big: using TPM
> warning: GenXPromoteArray: _ZTS8KernelIDILi1EE allocation size is too big: using TPM
> warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi2EEEE allocation size is too big: using TPM
> warning: GenXPromoteArray: _ZTS8KernelIDILi2EE allocation size is too big: using TPM
> warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi3EEEE allocation size is too big: using TPM
> warning: GenXPromoteArray: _ZTS8KernelIDILi3EE allocation size is too big: using TPM
> LLVM ERROR: Cannot find pointer replacement

Are these are known issues?

**Comment by bader** [comment]
> @kbobrovs, @DenisBakhvalov, I see failures in ESIMD/private_memory/pm_access_2.cpp and ESIMD/private_memory/pm_access_3.cpp after [comment update](https://github.com/intel/llvm/pull/3634/commits/f57217b002479646bef4e732d4f24cb8dd9a4e12):
> 
> > warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi1EEEE allocation size is too big: using TPM
> > warning: GenXPromoteArray: _ZTS8KernelIDILi1EE allocation size is too big: using TPM
> > warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi2EEEE allocation size is too big: using TPM
> > warning: GenXPromoteArray: _ZTS8KernelIDILi2EE allocation size is too big: using TPM
> > warning: GenXPromoteArray: _ZTSN2cl4sycl6detail19__pf_kernel_wrapperI8KernelIDILi3EEEE allocation size is too big: using TPM
> > warning: GenXPromoteArray: _ZTS8KernelIDILi3EE allocation size is too big: using TPM
> > LLVM ERROR: Cannot find pointer replacement
> 
> Are these are known issues?

I see these tests also fail in other PRs - https://github.com/intel/llvm/pull/3614, so these doesn't seem to be caused by changes from this PR.

File: clang/lib/Basic/Targets/SPIR.h
Comment (Naghasan) [reply clang/lib/Basic/Targets/SPIR.h]: >  the code doesn't need to be fixed, it's basically behaving as the implementation requires it

That's my interpretation as well. The change to `NOTE` make sense.

> Anastasia thinks Default should be mapped to OpenCL private address space

I fail to see any rational for that. SPIR is a target, it is not bound to an input language.

> I'm not sure if there is an issue with the SYCL spec.

Oh there is a lot of issues in the spec, but I don't think this is one of them. And I'm not fully convinced there is a clang design issue either: the underlying issue is we are trying to map the SYCL/CUDA/OpenCL memory model onto C/C++ which will have limitations (pointer escape for instances).

**Comment by Naghasan** [comment]
> Thanks!
> @AaronBallman, @Naghasan, does it look good to you?

LGTM :)

**Comment by bader** [comment]
> > Thanks!
> > @AaronBallman, @Naghasan, does it look good to you?
> 
> LGTM :)

Thanks a lot for review! I'll try to address your comments and remove the difference between https://github.com/llvm/llvm-project/commit/7818906ca134775edffb77857f436359d3a50b90 and what we landed to `sycl` branch in follow-up patches.

**Pull request merged**
